powerlaw/distributions.py

```diff
@@ -0,0 +1,2964 @@
+        following keys:
+
+        ``"type"``: The type of the constraint, either ``"eq"`` or
+        ``"ineq"``.
+
+        ``"fun"``": The function that implements the constraint.
+
+        ``"dists"``: The name of the distributions that this constraint
+        applies to. If not provided, constraint will be applied to
+        all distributions.
+
+        After some processsing and wrapping, these constraints are
+        eventually sent to ``scipy.optimize.minimize(constraints=...)``;
+        see their documentation for more information.
+
+    discrete_normalization : {"round", "sum"}, optional
+        Approximation method to use in calculating the PDF (especially the
+        PDF normalization constant) for a discrete distribution in the case
+        that there is no analytical expression available.
+
+        ``"round"`` uses the probability mass in the region ``[x - 0.5, x + 0.5]`` for each
+        data point ``x``.
+
+        ``"sum"`` simply sums the PDF over the defined range to compute the
+        normalization.
+
+    parent_Fit : Fit object, optional
+        A Fit object from which to use data, if it exists.
+
+    verbose : {0, 1, 2}, bool, optional
+        Whether to print debug and status information. ``0`` or ``False`` means
+        print no information (including no warnings), ``1`` or ``True`` means print
+        only warnings, and ``2`` means print warnings and status messages.
+
+    kwargs :
+        Parameter values for specific distributions can be passed as
+        keyword arguments.
+    """
+
+
+    # This is to indicate whether we should choose bounds based on
+    # xmax for the random number generation, or if this is handled
+    # within the _generate_random_continuous function. For most
+    # cases, if you can derive and implement the inverse CDF following
+    # the standard procedure, this should stay false. For an example
+    # of when you need it to be true, see Truncated power laws.
+    internally_bounded_rng = False
+
+    def __init__(self,
+                 data=None,
+                 xmin=None,
+                 xmax=None,
+                 discrete=False,
+                 fit_method='likelihood',
+                 parameters=None,
+                 parameter_ranges=None,
+                 parameter_constraints=None,
+                 discrete_normalization='round',
+                 parent_Fit=None,
+                 verbose=1,
+                 **kwargs):
+
+
+        # When defining a subclass of this one, you should define this
+        # list as a property of the class
+        #parameter_names = []
+        # You also will need to define this in the subclass with the
+        # upper and lower bound for each parameter.
+        #DEFAULT_PARAMETER_RANGES = {}
+        # (but we don't define them here because we don't want to overwrite
+        # the values created in the subclass)
+        # And the name of the distribution
+        #name = 'name'
```

Comment:
these might be definable using ABCMeta methods, though I'm not super familiar with those so I'm not requesting a change, just noting the possibility.